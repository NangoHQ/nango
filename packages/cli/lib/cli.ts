import fs from 'node:fs';
import path, { dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import chalk from 'chalk';
import chokidar from 'chokidar';
import ejs from 'ejs';
import * as dotenv from 'dotenv';

import { getNangoRootPath, printDebug } from './utils.js';
import { loadYamlAndGenerate } from './services/model.service.js';
import { TYPES_FILE_NAME, exampleSyncName } from './constants.js';
import { compileAllFiles, compileSingleFile, getFileToCompile } from './services/compile.service.js';
import { getLayoutMode } from './utils/layoutMode.js';
import { getProviderConfigurationFromPath, nangoConfigFile } from '@nangohq/nango-yaml';
import { NANGO_VERSION } from './version.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

dotenv.config();

export const version = (debug: boolean) => {
    if (debug) {
        printDebug('Looking up the version first for a local path first then globally');
    }
    const version = NANGO_VERSION;

    console.log(chalk.green('Nango CLI version:'), version);
};

export function generate({ fullPath, debug = false }: { fullPath: string; debug?: boolean }) {
    const syncTemplateContents = fs.readFileSync(path.resolve(__dirname, './templates/sync.ejs'), 'utf8');
    const actionTemplateContents = fs.readFileSync(path.resolve(__dirname, './templates/action.ejs'), 'utf8');
    const githubExampleTemplateContents = fs.readFileSync(path.resolve(__dirname, './templates/github.sync.ejs'), 'utf8');
    const onEventTemplateContents = fs.readFileSync(path.resolve(__dirname, './templates/on-event.ejs'), 'utf8');

    const parsed = loadYamlAndGenerate({ fullPath, debug });
    if (!parsed) {
        return;
    }

    const allSyncNames: Record<string, boolean> = {};

    for (const integration of parsed.integrations) {
        const { syncs, actions, onEventScripts, providerConfigKey } = integration;

        if (onEventScripts) {
            const type = 'on-event';
            for (const name of Object.values(onEventScripts).flat()) {
                const rendered = ejs.render(onEventTemplateContents, {
                    interfaceFileName: TYPES_FILE_NAME.replace('.ts', '')
                });
                const stripped = rendered.replace(/^\s+/, '');

                if (!fs.existsSync(path.resolve(fullPath, `${providerConfigKey}/${type}s/${name}.ts`))) {
                    fs.mkdirSync(path.resolve(fullPath, `${providerConfigKey}/${type}s`), { recursive: true });
                    fs.writeFileSync(path.resolve(fullPath, `${providerConfigKey}/${type}s/${name}.ts`), stripped);
                    if (debug) {
                        printDebug(`Created ${name}.ts file`);
                    }
                } else {
                    if (debug) {
                        printDebug(`${name}.ts file already exists, so will not overwrite it.`);
                    }
                }
            }
        }

        for (const flow of [...syncs, ...actions]) {
            const { name, type, output, input } = flow;
            const layoutMode = getLayoutMode({ fullPath, providerConfigKey, scriptName: name, type });
            const uniqueName = layoutMode === 'root' ? name : `${providerConfigKey}-${name}`;

            if (allSyncNames[uniqueName] === undefined) {
                // a sync and an action within the same provider cannot have the same name
                allSyncNames[uniqueName] = true;
            } else {
                console.log(chalk.red(`The ${type} name ${name} is duplicated in the ${nangoConfigFile} file. All sync and action names must be unique.`));
                process.exit(1);
            }

            if (fs.existsSync(path.resolve(fullPath, `${name}.ts`)) || fs.existsSync(path.resolve(fullPath, `${providerConfigKey}/${type}s/${name}.ts`))) {
                if (debug) {
                    printDebug(`${name}.ts file already exists, so will not overwrite it.`);
                }
                continue;
            }

            if (debug) {
                printDebug(`Generating ${name} integration in layout mode ${layoutMode}`);
            }

            const flowNameCamel = name
                .split('-')
                .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
                .join('');

            let ejsTemplateContents = '';

            if (name === exampleSyncName && type === 'sync') {
                ejsTemplateContents = githubExampleTemplateContents;
            } else {
                ejsTemplateContents = type === 'sync' ? syncTemplateContents : actionTemplateContents;
            }

            const models = (output ? [...output, input] : [input]).filter(Boolean);
            const rendered = ejs.render(ejsTemplateContents, {
                syncName: flowNameCamel,
                interfacePath: layoutMode === 'root' ? './' : '../../',
                interfaceFileName: TYPES_FILE_NAME.replace('.ts', ''),
                output: output && output.length > 0 ? output.join(' | ') : null,
                input: input,
                modelNames: models.join(', '),
                hasWebhook: type === 'sync' && flow.webhookSubscriptions && flow.webhookSubscriptions.length > 0
            });

            const stripped = rendered.replace(/^\s+/, '');

            if (layoutMode === 'root') {
                fs.writeFileSync(path.resolve(fullPath, `${name}.ts`), stripped);
            } else {
                fs.mkdirSync(path.resolve(fullPath, `${providerConfigKey}/${type}s`), { recursive: true });
                fs.writeFileSync(path.resolve(fullPath, `${providerConfigKey}/${type}s/${name}.ts`), stripped);
            }
            if (debug) {
                console.log(chalk.green(`Created ${name}.ts file`));
            }
        }
    }
}

export function tscWatch({ fullPath, debug = false }: { fullPath: string; debug?: boolean }) {
    const tsconfig = fs.readFileSync(path.resolve(getNangoRootPath(), 'tsconfig.dev.json'), 'utf8');
    const parsed = loadYamlAndGenerate({ fullPath, debug });
    if (!parsed) {
        return;
    }

    const watchPath = ['./**/*.ts', `./${nangoConfigFile}`];

    if (debug) {
        printDebug(`Watching ${watchPath.join(', ')}`);
    }

    const watcher = chokidar.watch(watchPath, {
        ignoreInitial: false,
        ignored: (filePath: string) => {
            const relativePath = path.relative(__dirname, filePath);
            return relativePath.includes('node_modules') || path.basename(filePath) === TYPES_FILE_NAME;
        }
    });

    const distDir = path.join(fullPath, 'dist');

    if (!fs.existsSync(distDir)) {
        if (debug) {
            printDebug(`Creating ${distDir} directory`);
        }
        fs.mkdirSync(distDir);
    }

    watcher.on('add', async (filePath: string) => {
        if (filePath === nangoConfigFile) {
            return;
        }
        await compileSingleFile({ fullPath, file: getFileToCompile({ fullPath, filePath }), tsconfig, parsed, debug });
    });

    watcher.on('unlink', (filePath: string) => {
        if (filePath === nangoConfigFile) {
            return;
        }
        const providerConfiguration = getProviderConfigurationFromPath({ filePath, parsed });
        const baseName = path.basename(filePath, '.ts');
        const fileName = providerConfiguration ? `${baseName}-${providerConfiguration.providerConfigKey}.js` : `${baseName}.js`;
        const jsFilePath = `./dist/${fileName}`;

        try {
            fs.unlinkSync(jsFilePath);
        } catch {
            console.log(chalk.red(`Error deleting ${jsFilePath}`));
        }
    });

    watcher.on('change', async (filePath: string) => {
        if (filePath === nangoConfigFile) {
            await compileAllFiles({ fullPath, debug });
            return;
        }
        await compileSingleFile({ fullPath, file: getFileToCompile({ fullPath, filePath }), parsed, debug });
    });
}

export function configWatch({ fullPath, debug = false }: { fullPath: string; debug?: boolean }) {
    const watchPath = path.join(fullPath, nangoConfigFile);
    if (debug) {
        printDebug(`Watching ${watchPath}`);
    }
    const watcher = chokidar.watch(watchPath, { ignoreInitial: true });

    watcher.on('change', () => {
        loadYamlAndGenerate({ fullPath, debug });
    });
}
