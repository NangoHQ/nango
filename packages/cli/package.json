{
    "name": "nango",
    "version": "0.53.1",
    "description": "Nango's CLI tool.",
    "type": "module",
    "main": "dist/index.js",
    "bin": {
        "nango": "dist/index.js"
    },
    "types": "dist/nango-sync.d.ts",
    "keywords": [],
    "author": "bastien@nango.dev",
    "repository": {
        "type": "git",
        "url": "git+https://github.com/NangoHQ/nango.git",
        "directory": "packages/cli"
    },
    "license": "SEE LICENSE IN LICENSE FILE IN GIT REPOSITORY",
    "scripts": {
        "copy:files": "copyfiles -u 1 lib/templates/* dist",
        "prepublishOnly": "npm run copy:files"
    },
    "dependencies": {
        "@babel/parser": "^7.22.5",
        "@babel/traverse": "^7.22.5",
        "@babel/types": "^7.22.5",
        "@nangohq/nango-yaml": "0.53.1",
        "@nangohq/node": "0.53.1",
        "@nangohq/runner-sdk": "0.53.1",
        "@swc/core": "^1.5.25",
        "@types/unzipper": "^0.10.10",
        "ajv": "^8.17.1",
        "ajv-errors": "^3.0.0",
        "axios": "^1.7.9",
        "chalk": "^5.3.0",
        "chokidar": "^3.5.3",
        "commander": "^10.0.1",
        "conf": "12.0.0",
        "copyfiles": "^2.4.1",
        "dotenv": "^16.0.3",
        "ejs": "^3.1.10",
        "figlet": "^1.6.0",
        "glob": "^10.3.10",
        "import-meta-resolve": "^4.1.0",
        "js-yaml": "^4.1.0",
        "npm-package-arg": "^10.1.0",
        "promptly": "^3.2.0",
        "semver": "^7.5.4",
        "serialize-error": "11.0.3",
        "ts-json-schema-generator": "^2.3.0",
        "ts-node": "^10.9.1",
        "tsup": "^8.1.0",
        "typescript": "5.7.3",
        "unzipper": "^0.12.3",
        "zod": "3.24.1"
    },
    "devDependencies": {
        "@babel/core": "^7.22.1",
        "@babel/preset-env": "^7.22.4",
        "@babel/preset-typescript": "^7.21.5",
        "@nangohq/types": "0.53.1",
        "@types/babel__traverse": "^7.20.1",
        "@types/babel-traverse": "^6.25.7",
        "@types/commander": "^2.12.2",
        "@types/ejs": "^3.1.2",
        "@types/figlet": "^1.5.6",
        "@types/glob": "^8.1.0",
        "@types/json-schema": "7.0.15",
        "@types/node": "^20.1.4",
        "@types/npm-package-arg": "^6.1.1",
        "@types/promptly": "^3.0.2",
        "babel-loader": "^9.1.2",
        "esbuild": "^0.25.0",
        "json-schema": "0.4.0",
        "strip-ansi": "7.1.0",
        "vitest": "2.1.9"
    },
    "engines": {
        "node": ">=18.0"
    },
    "files": [
        "dist/**/*",
        "docker/docker-compose.yaml",
        "lib/nango.yaml.schema.v1.json",
        "lib/nango.yaml.schema.v2.json",
        "scripts/v1-v2.js",
        "tsconfig.dev.json",
        "!**/*.json",
        "README.md"
    ]
}
