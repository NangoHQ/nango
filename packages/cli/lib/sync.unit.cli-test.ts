import fs from 'node:fs';
import os from 'node:os';
import path, { join } from 'node:path';

import yaml from 'js-yaml';
import stripAnsi from 'strip-ansi';
import { afterEach, describe, expect, it, vi } from 'vitest';

import { generate } from './cli.js';
import { compileAllFiles, compileSingleFile, getFileToCompile } from './services/compile.service.js';
import { parse } from './services/config.service.js';
import { init } from './services/init.service.js';
import { directoryMigration, endpointMigration } from './services/migration.service.js';
import parserService from './services/parser.service.js';
import { copyDirectoryAndContents, fixturesPath, getTestDirectory } from './tests/helpers.js';

describe('generate function tests', () => {
    // Not the best but until we have a logger it will work
    const consoleMock = vi.spyOn(console, 'log').mockImplementation(() => undefined);

    afterEach(() => {
        consoleMock.mockReset();
    });

    it('should generate a different sync correctly', async () => {
        const dir = await getTestDirectory('generate');
        init({ absolutePath: dir });
        const data = {
            integrations: {
                'demo-github-integration': {
                    'some-other-sync': {
                        type: 'sync',
                        runs: 'every half hour',
                        returns: ['GithubIssue']
                    }
                }
            },
            models: {
                GithubIssue: {
                    id: 'integer',
                    owner: 'string',
                    repo: 'string',
                    issue_number: 'number',
                    title: 'string',
                    author: 'string',
                    author_id: 'string',
                    state: 'string',
                    date_created: 'date',
                    date_last_modified: 'date',
                    body: 'string'
                }
            }
        };
        const yamlData = yaml.dump(data);
        await fs.promises.writeFile(join(dir, 'nango.yaml'), yamlData, 'utf8');
        generate({ debug: true, fullPath: dir });
        expect(fs.existsSync(join(dir, 'demo-github-integration/syncs/some-other-sync.ts'))).toBe(true);
    });

    it('should support a single model return in v1 format', async () => {
        const dir = await getTestDirectory('single-model-v1');
        init({ absolutePath: dir });
        const data = {
            integrations: {
                'demo-github-integration': {
                    'single-model-return': {
                        type: 'sync',
                        runs: 'every half hour',
                        returns: 'GithubIssue'
                    }
                }
            },
            models: {
                GithubIssue: {
                    id: 'integer',
                    owner: 'string',
                    repo: 'string',
                    issue_number: 'number',
                    title: 'string',
                    author: 'string',
                    author_id: 'string',
                    state: 'string',
                    date_created: 'date',
                    date_last_modified: 'date',
                    body: 'string'
                }
            }
        };
        const yamlData = yaml.dump(data);
        await fs.promises.writeFile(join(dir, 'nango.yaml'), yamlData, 'utf8');
        generate({ debug: false, fullPath: dir });
        expect(fs.existsSync(join(dir, 'demo-github-integration/syncs/single-model-return.ts'))).toBe(true);
    });

    it('should support a single model return in v2 format', async () => {
        const dir = await getTestDirectory('single-model-v2');
        init({ absolutePath: dir });
        const data = {
            integrations: {
                'demo-github-integration': {
                    syncs: {
                        'single-model-return': {
                            runs: 'every half hour',
                            endpoint: 'GET /issues',
                            output: 'GithubIssue'
                        }
                    }
                }
            },
            models: {
                GithubIssue: {
                    id: 'integer',
                    owner: 'string',
                    repo: 'string',
                    issue_number: 'number',
                    title: 'string',
                    author: 'string',
                    author_id: 'string',
                    state: 'string',
                    date_created: 'date',
                    date_last_modified: 'date',
                    body: 'string'
                }
            }
        };
        const yamlData = yaml.dump(data);
        await fs.promises.writeFile(join(dir, 'nango.yaml'), yamlData, 'utf8');
        generate({ debug: false, fullPath: dir });
        expect(fs.existsSync(join(dir, 'demo-github-integration/syncs/single-model-return.ts'))).toBe(true);
    });

    it('should not create a file if endpoint is missing from a v2 config', async () => {
        const dir = await getTestDirectory('endpoint-missing-v2');
        init({ absolutePath: dir });
        const data = {
            integrations: {
                'demo-github-integration': {
                    syncs: {
                        'single-model-return': {
                            type: 'sync',
                            runs: 'every half hour',
                            returns: 'GithubIssue'
                        }
                    }
                }
            },
            models: {
                GithubIssue: {
                    id: 'integer',
                    owner: 'string',
                    repo: 'string',
                    issue_number: 'number',
                    title: 'string',
                    author: 'string',
                    author_id: 'string',
                    state: 'string',
                    date_created: 'date',
                    date_last_modified: 'date',
                    body: 'string'
                }
            }
        };
        const yamlData = yaml.dump(data);
        await fs.promises.writeFile(join(dir, 'nango.yaml'), yamlData, 'utf8');
        expect(fs.existsSync(join(dir, 'demo-github-integration/syncs/single-model-return.ts'))).toBe(false);
    });

    it('should generate missing from a v2 config', async () => {
        const dir = await getTestDirectory('v2-incremental-compile');
        init({ absolutePath: dir });
        const data = {
            integrations: {
                'demo-github-integration': {
                    syncs: {
                        'single-model-issue-output': {
                            runs: 'every half hour',
                            endpoint: 'GET /tickets/issue',
                            output: 'GithubIssue'
                        }
                    }
                }
            },
            models: {
                GithubIssue: {
                    id: 'integer',
                    owner: 'string',
                    repo: 'string',
                    issue_number: 'number',
                    title: 'string',
                    author: 'string',
                    author_id: 'string',
                    state: 'string',
                    date_created: 'date',
                    date_last_modified: 'date',
                    body: 'string'
                }
            }
        };
        const yamlData = yaml.dump(data);
        await fs.promises.writeFile(join(dir, 'nango.yaml'), yamlData, 'utf8');
        generate({ debug: false, fullPath: dir });
        expect(fs.existsSync(join(dir, 'demo-github-integration/syncs/single-model-issue-output.ts'))).toBe(true);
    });

    it('should throw an error if a model is missing an id that is actively used', async () => {
        const dir = await getTestDirectory('missing-id');
        init({ absolutePath: dir });
        const data = {
            integrations: {
                'demo-github-integration': {
                    'single-model-return': {
                        type: 'sync',
                        runs: 'every half hour',
                        returns: 'GithubIssue'
                    }
                }
            },
            models: {
                GithubIssue: {
                    owner: 'string',
                    repo: 'string',
                    issue_number: 'number',
                    title: 'string',
                    author: 'string',
                    author_id: 'string',
                    state: 'string',
                    date_created: 'date',
                    date_last_modified: 'date',
                    body: 'string'
                }
            }
        };

        const yamlData = yaml.dump(data);
        await fs.promises.writeFile(join(dir, 'nango.yaml'), yamlData, 'utf8');

        const acc: string[] = [];
        consoleMock.mockImplementation((m) => acc.push(typeof m === 'string' ? stripAnsi(m) : m));
        generate({ debug: false, fullPath: dir });

        expect(acc).toStrictEqual([
            'demo-github-integration > syncs > single-model-return > [output]',
            '  error Model "GithubIssue" doesn\'t have an id field. This is required to be able to uniquely identify the data record [model_missing_id]',
            '',
            'Your nango.yaml contains some errors'
        ]);
    });

    it('should allow models to end with an "s"', async () => {
        const dir = await getTestDirectory('model-with-an-s');
        init({ absolutePath: dir });
        const data = {
            integrations: {
                'demo-github-integration': {
                    'single-model-return': {
                        type: 'sync',
                        runs: 'every half hour',
                        returns: 'GithubIssues'
                    }
                }
            },
            models: {
                GithubIssues: {
                    id: 'string',
                    owner: 'string',
                    repo: 'string',
                    issue_number: 'number',
                    title: 'string',
                    author: 'string',
                    author_id: 'string',
                    state: 'string',
                    date_created: 'date',
                    date_last_modified: 'date',
                    body: 'string'
                }
            }
        };
        const yamlData = yaml.dump(data);
        await fs.promises.writeFile(join(dir, 'nango.yaml'), yamlData, 'utf8');
        generate({ debug: false, fullPath: dir });
        const modelsFile = await fs.promises.readFile(join(dir, 'models.ts'), 'utf8');
        expect(modelsFile).toContain('export interface GithubIssues');
    });

    it('should not throw an error if a model is missing an id for an action', async () => {
        const dir = await getTestDirectory('missing-id-action');
        init({ absolutePath: dir });
        const data = {
            integrations: {
                'demo-github-integration': {
                    'single-model-return': {
                        type: 'action',
                        returns: 'GithubIssue'
                    }
                }
            },
            models: {
                GithubIssue: {
                    owner: 'string',
                    repo: 'string',
                    issue_number: 'number',
                    title: 'string',
                    author: 'string',
                    author_id: 'string',
                    state: 'string',
                    date_created: 'date',
                    date_last_modified: 'date',
                    body: 'string'
                }
            }
        };

        const yamlData = yaml.dump(data);
        await fs.promises.writeFile(join(dir, 'nango.yaml'), yamlData, 'utf8');
        generate({ debug: false, fullPath: dir });
    });

    it('should allow javascript primitives as a return type with no model', async () => {
        const dir = await getTestDirectory('model-returns-primitives');
        init({ absolutePath: dir });
        const data = {
            integrations: {
                'demo-github-integration': {
                    'single-model-return': {
                        type: 'sync',
                        returns: 'string'
                    },
                    'single-model-return-action': {
                        type: 'action',
                        returns: 'string'
                    }
                }
            }
        };
        const yamlData = yaml.dump(data);
        await fs.promises.writeFile(join(dir, 'nango.yaml'), yamlData, 'utf8');
        generate({ debug: false, fullPath: dir });
    });

    it('should catch non javascript primitives in the config', async () => {
        const dir = await getTestDirectory('model-returns-invalid-primitives');
        init({ absolutePath: dir });
        const data = {
            integrations: {
                'demo-github-integration': {
                    'single-model-return': {
                        type: 'sync',
                        returns: 'string'
                    },
                    'single-model-return-action': {
                        type: 'action',
                        returns: 'strings'
                    }
                }
            }
        };
        const yamlData = yaml.dump(data);
        await fs.promises.writeFile(join(dir, 'nango.yaml'), yamlData, 'utf8');
        expect(generate({ debug: false, fullPath: dir })).toBeUndefined();
    });

    it('should not complain of try catch not being awaited', () => {
        const awaiting = parserService.callsAreUsedCorrectly(join(fixturesPath, 'sync.ts'), 'sync', ['GithubIssue']);
        expect(awaiting).toBe(true);
    });

    it('should complain when a return statement is used', () => {
        const noReturnUsed = parserService.callsAreUsedCorrectly(join(fixturesPath, 'return-sync.ts'), 'sync', ['GithubIssue']);
        expect(noReturnUsed).toBe(false);
    });

    it('should not complain when a return statement is used but does not return anything', () => {
        const noReturnUsed = parserService.callsAreUsedCorrectly(join(fixturesPath, 'void-return-sync.ts'), 'sync', ['GithubIssue']);
        expect(noReturnUsed).toBe(true);
    });

    it('should not complain when a return statement is used in a nested function', () => {
        const noReturnUsed = parserService.callsAreUsedCorrectly(join(fixturesPath, 'nested-return-sync.ts'), 'sync', ['GreenhouseEeoc']);
        expect(noReturnUsed).toBe(true);
    });

    it('should complain of a non try catch not being awaited', () => {
        const awaiting = parserService.callsAreUsedCorrectly(join(fixturesPath, 'failing-sync.ts'), 'sync', ['GithubIssue']);
        expect(awaiting).toBe(false);
    });

    it('should not complain about a correct model', () => {
        const usedCorrectly = parserService.callsAreUsedCorrectly(join(fixturesPath, 'bad-model.ts'), 'sync', ['SomeBadModel']);
        expect(usedCorrectly).toBe(true);
    });

    it('should not complain about awaiting when it is returned for an action', () => {
        const awaiting = parserService.callsAreUsedCorrectly(join(fixturesPath, 'no-async-return.ts'), 'action', ['SomeModel']);
        expect(awaiting).toBe(true);
    });

    it('should complain about an incorrect model', () => {
        const awaiting = parserService.callsAreUsedCorrectly(join(fixturesPath, 'bad-model.ts'), 'sync', ['GithubIssue']);
        expect(awaiting).toBe(false);
    });

    it('should complain if retryOn is used without retries', () => {
        const usedCorrectly = parserService.callsAreUsedCorrectly(join(fixturesPath, 'retry-on-bad.ts'), 'sync', ['GithubIssue']);
        expect(usedCorrectly).toBe(false);
    });

    it('should not complain if retryOn is used with retries', () => {
        const usedCorrectly = parserService.callsAreUsedCorrectly(join(fixturesPath, 'retry-on-good.ts'), 'sync', ['GithubIssue']);
        expect(usedCorrectly).toBe(false);
    });

    it('should be able to compile files in nested directories', async () => {
        const dir = await getTestDirectory('nested');
        init({ absolutePath: dir });

        await copyDirectoryAndContents(join(fixturesPath, 'nango-yaml/v2/nested-integrations/hubspot'), join(dir, 'hubspot'));
        await copyDirectoryAndContents(join(fixturesPath, 'nango-yaml/v2/nested-integrations/github'), join(dir, 'github'));
        await fs.promises.copyFile(join(fixturesPath, 'nango-yaml/v2/nested-integrations/nango.yaml'), join(dir, 'nango.yaml'));

        {
            // Compile everything
            const success = await compileAllFiles({ fullPath: dir, debug: true });

            //. these should report any failed paths somehow, not just true!=false
            expect(fs.existsSync(join(dir, 'models.ts'))).toBe(true);
            expect(fs.existsSync(join(dir, 'hubspot/syncs/contacts.ts'))).toBe(true);
            expect(fs.existsSync(join(dir, 'dist/contacts-hubspot.js'))).toBe(true);
            expect(fs.existsSync(join(dir, 'dist/issues-github.js'))).toBe(true);

            expect(success).toBe(true);
        }

        {
            // Compile one file
            const success = await compileAllFiles({ fullPath: dir, debug: true, scriptName: 'contacts', providerConfigKey: 'hubspot', type: 'syncs' });
            expect(success).toBe(true);
        }
    });

    it('should be backwards compatible with the single directory for integration files', async () => {
        const dir = await getTestDirectory('old-directory');
        init({ absolutePath: dir });

        await copyDirectoryAndContents(join(fixturesPath, 'nango-yaml/v2/non-nested-integrations'), dir);

        const success = await compileAllFiles({ fullPath: dir, debug: false });

        expect(fs.existsSync(join(dir, 'models.ts'))).toBe(true);
        expect(fs.existsSync(join(dir, 'contacts.ts'))).toBe(true);
        expect(fs.existsSync(join(dir, 'dist/contacts-hubspot.js'))).toBe(true);
        expect(success).toBe(true);
    });

    it('should be able to migrate-to-directories', async () => {
        const dir = await getTestDirectory('old-directory-migrate');
        init({ absolutePath: dir });

        await copyDirectoryAndContents(join(fixturesPath, 'nango-yaml/v2/non-nested-integrations'), dir);

        await directoryMigration(dir);
        expect(fs.existsSync(join(dir, 'hubspot/syncs/contacts.ts'))).toBe(true);
        expect(fs.existsSync(join(dir, 'hubspot/actions/create-contact.ts'))).toBe(true);
        expect(fs.existsSync(join(dir, 'contacts.ts'))).toBe(false);
        expect(fs.existsSync(join(dir, 'create-contacts.ts'))).toBe(false);

        const success = await compileAllFiles({ fullPath: dir, debug: false });
        expect(fs.existsSync(join(dir, 'models.ts'))).toBe(true);
        expect(fs.existsSync(join(dir, 'dist/contacts-hubspot.js'))).toBe(true);

        expect(success).toBe(true);
    });

    it('should be able to compile and run imported files', async () => {
        const dir = await getTestDirectory('relative-imports');
        init({ absolutePath: dir });

        await copyDirectoryAndContents(join(fixturesPath, 'nango-yaml/v2/relative-imports/github'), join(dir, 'github'));
        await fs.promises.copyFile(join(fixturesPath, 'nango-yaml/v2/relative-imports/nango.yaml'), join(dir, 'nango.yaml'));

        const success = await compileAllFiles({ fullPath: dir, debug: false });

        const module = await import(path.normalize(join(dir, 'dist/issues-github.js')));

        const result = module.default.default();
        expect(result).toBe('Hello, world!');

        expect(success).toBe(true);
    });

    it('should compile helper functions and throw an error if there is a complication error with an imported file', async () => {
        const name = 'relative-imports-with-error';
        const dir = await getTestDirectory('relative-imports-with-error');
        init({ absolutePath: dir });

        await copyDirectoryAndContents(join(fixturesPath, `nango-yaml/v2/${name}/github`), join(dir, 'github'));
        await fs.promises.copyFile(join(fixturesPath, `nango-yaml/v2/${name}/nango.yaml`), join(dir, 'nango.yaml'));

        const parsing = parse(path.resolve(join(fixturesPath, `nango-yaml/v2/${name}`)));
        if (parsing.isErr()) {
            throw parsing.error;
        }

        const result = await compileSingleFile({
            fullPath: dir,
            file: getFileToCompile({ fullPath: dir, filePath: join(dir, './github/actions/gh-issues.ts') }),
            parsed: parsing.value.parsed!,
            debug: false
        });
        expect(result).toBe(false);
    });

    it('should complain if a nango call is used incorrectly in a nested file', async () => {
        const name = 'relative-imports-with-nango-misuse';
        const dir = await getTestDirectory('relative-imports-with-nango-misuse');
        init({ absolutePath: dir });

        await copyDirectoryAndContents(join(fixturesPath, `nango-yaml/v2/${name}/github`), join(dir, 'github'));
        await fs.promises.copyFile(join(fixturesPath, `nango-yaml/v2/${name}/nango.yaml`), join(dir, 'nango.yaml'));

        const parsing = parse(path.resolve(join(fixturesPath, `nango-yaml/v2/${name}`)));
        if (parsing.isErr()) {
            throw parsing.error;
        }

        const result = await compileSingleFile({
            fullPath: dir,
            file: getFileToCompile({ fullPath: dir, filePath: join(dir, './github/actions/gh-issues.ts') }),
            parsed: parsing.value.parsed!,
            debug: false
        });
        expect(result).toBe(false);
    });

    it('should not allow imports higher than the current directory', async () => {
        const name = 'relative-imports-with-higher-import';
        const dir = await getTestDirectory('relative-imports-with-higher-import');
        init({ absolutePath: dir });

        await copyDirectoryAndContents(join(fixturesPath, `nango-yaml/v2/${name}/github`), join(dir, 'github'));
        await fs.promises.copyFile(join(fixturesPath, `nango-yaml/v2/${name}/nango.yaml`), join(dir, 'nango.yaml'));
        await fs.promises.copyFile(join(fixturesPath, `nango-yaml/v2/${name}/github/actions/welcomer.ts`), join(dir, 'welcomer.ts'));

        const parsing = parse(path.resolve(join(fixturesPath, `nango-yaml/v2/${name}`)));
        if (parsing.isErr()) {
            throw parsing.error;
        }

        const result = await compileSingleFile({
            fullPath: dir,
            file: getFileToCompile({ fullPath: dir, filePath: join(dir, './github/actions/gh-issues.ts') }),
            parsed: parsing.value.parsed!,
            debug: false
        });
        expect(result).toBe(false);
    });

    // Problem with double lines
    it.skipIf(os.platform() === 'win32')('should be able to migrate-endpoints', async () => {
        const dir = await getTestDirectory('old-endpoint');
        init({ absolutePath: dir });

        const dest = join(dir, 'nango.yaml');
        await fs.promises.copyFile(join(fixturesPath, 'nango-yaml/v2/nango.yaml'), dest);

        endpointMigration(dir);

        const content = await fs.promises.readFile(dest, 'utf8');

        expect(content).toMatchSnapshot();
    });

    // Windows symlink are annoying to create
    it.skipIf(os.platform() === 'win32')('should be able to compile files in symlink', async () => {
        const dir = await getTestDirectory('symlink');
        init({ absolutePath: dir });

        await fs.promises.rm(join(dir, 'nango.yaml'));
        await fs.promises.symlink(join(fixturesPath, 'nango-yaml/v2/nested-integrations/nango.yaml'), join(dir, 'nango.yaml'));
        await fs.promises.symlink(join(fixturesPath, 'nango-yaml/v2/nested-integrations/github'), join(dir, 'github'));
        await fs.promises.symlink(join(fixturesPath, 'nango-yaml/v2/nested-integrations/hubspot'), join(dir, 'hubspot'));

        {
            // Compile everything
            const success = await compileAllFiles({ fullPath: dir, debug: true });
            expect(fs.existsSync(join(dir, 'models.ts'))).toBe(true);
            expect(fs.existsSync(join(dir, 'hubspot/syncs/contacts.ts'))).toBe(true);
            expect(fs.existsSync(join(dir, 'dist/contacts-hubspot.js'))).toBe(true);
            expect(fs.existsSync(join(dir, 'dist/issues-github.js'))).toBe(true);
            expect(success).toBe(true);
        }

        {
            // Compile one file
            const success = await compileAllFiles({ fullPath: dir, debug: true, scriptName: 'contacts', providerConfigKey: 'hubspot', type: 'syncs' });
            expect(success).toBe(true);
        }
    });
});
