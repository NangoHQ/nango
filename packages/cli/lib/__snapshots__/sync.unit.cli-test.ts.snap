// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate function tests > should be able to migrate-endpoints 1`] = `
"integrations:
    demo-github-integration:
        syncs:
            github-issue-example:
                description: |
                    Sync github issues continuously from public repos
                sync_type: incremental
                auto_start: true
                runs: every half hour
                scopes: public_repo
                output: GithubIssue
                endpoint:
                  method: GET
                  path: /github/issues
                webhook-subscriptions:
                    - issue.creation
            github-issue-example-two:
                description: |
                    Sync github issues continuously from public repos example two
                sync_type: incremental
                auto_start: true
                runs: every hour
                endpoint:
                  method: GET
                  path: /ticketing/tickets-two
                scopes: public_repo
                output: GithubIssue2
            github-multiple-models:
                description: Sync github issues to multiple models
                sync_type: full
                auto_start: true
                runs: every 5 minutes
                endpoint:
                    - method: GET
                      path: /ticketing/ticket
                    - method: GET
                      path: /ticketing/pr
                output:
                    - GithubIssue3
                    - GithubPR
        actions:
            github-get-issue:
                description: Get a GitHub issue.
                endpoint:
                  method: GET
                  path: /ticketing/tickets/{GithubCreateIssueInput:id}
                output: GithubIssueAction
                scopes: repo:read
            github-create-issue:
                description: Creates a GitHub issue.
                endpoint:
                  method: POST
                  path: /ticketing/tickets
                scopes: repo:write
                input: GithubCreateIssueInput
                output: GithubCreateOutput
            github-delete-issue:
                description: Deletes a GitHub issue.
                endpoint:
                  method: DELETE
                  path: /ticketing/tickets/{GithubIssue:id}
                scopes: repo:write
                output: boolean

models:
    GithubIssue:
        id: integer
        owner: string
        repo: string
        issue_number: number
        title: string
        author: string
        author_id: string
        state: string
        date_created: date
        date_last_modified: date
        body: string
    GithubIssueAction:
        __extends: GithubIssue
    GithubIssue2:
        __extends: GithubIssue
    GithubIssue3:
        __extends: GithubIssue
    GithubCreateIssueInput:
        __extends: GithubIssue
    GithubCreateOutput:
        result: GithubIssue
    GithubPR:
        __extends: GithubIssue
"
`;
